src/core/protocol-authorization.ts

```diff
@@ -52,6 +48,47 @@ export class ProtocolAuthorization {
       ancestorMessageChain,
     );
 
+    // get the rule set for the inbound message
+    const inboundMessageRuleSet = ProtocolAuthorization.getRuleSet(
+      incomingMessage,
+      protocolDefinition,
+    );
+
+    // If the incoming message is writing a $globalRole record, validate that the recipient is unique
+    await ProtocolAuthorization.verifyUniqueRoleRecipient(
+      tenant,
+      incomingMessage,
+      inboundMessageRuleSet,
+      messageStore,
+    );
+
+    // verify allowed condition of incoming message
+    await ProtocolAuthorization.verifyActionCondition(tenant, incomingMessage, messageStore);
```

Comment:
Could be arguable, but this feels more like actual authorization than "structure validation", I understand the code is "less clean" if it is `authorization` path because of the `RecordsRead` check that you removed. But I am okay with that tradeoff.